1주차 진행 - My Diary

yhyojoo opened the pull request:
## 이번 주가 지나면 사용자는 무엇을 할 수 있나요?
다이어리와 할 일을 추가, 수정, 삭제할 수 있습니다. 
추가된 다이어리와 힐 일에 대한 상세 조회, 목록 조회도 가능합니다.

## 이번 주에 구현할 기능을 우선순위대로 작성해 주세요.

- 개발 환경 세팅
- 다이어리 목록 조회 기능
- 다이어리 상세 조회 기능 
- 다이어리 추가 기능
- 다이어리 수정 기능
- 다이어리 삭제 기능
- 할 일 목록 조회 기능
- 할 일 상세 조회 기능 
- 할 일 추가 기능
- 할 일 수정 기능
- 할 일 삭제 기능

yhyojoo committed:
개발 환경을 세팅하라

yhyojoo committed:
다이어리 도메인 객체 클래스를 생성하라

yhyojoo committed:
다이어리 객체 빌더 테스트를 추가하라

yhyojoo committed:
개행 문자를 추가하라

yhyojoo committed:
다이어리 정보 수정 테스트를 추가하라

yhyojoo committed:
다이어리 정보 수정 기능을 구현하라

yhyojoo committed:
다이어리 목록 조회 기능을 구현하라

yhyojoo committed:
다이어리 목록 조회 테스트를 작성하라

yhyojoo committed:
다이어리 Repository 인터페이스를 생성하고, CRUD 메소드를 선언하라

yhyojoo committed:
다이어리 조회 테스트를 추가하라

yhyojoo committed:
다이어리 조회 기능을 구현하라

yhyojoo committed:
다이어리를 찾을 수 없는 예외를 처리하라

johngrib commented on src/test/java/com/codesoom/project/domain/DiaryTest.java:
```diff
@@ -29,4 +29,4 @@ void creationWithBuilder() {
     }
-}
\ No newline at end of file
+}
```
이런 작업은 e5dc602 에 합쳐줬어도 좋았을 거에요.

johngrib commented on src/test/java/com/codesoom/project/domain/DiaryTest.java:
```diff
@@ -29,4 +30,15 @@ void creationWithBuilder() {
+    @Test
+    void updateWith() {
+        diary.updateWith(Diary.builder()
+                .title("오늘의 다이어리")
+                .comment("보람찬 하루였다")
+                .build());
+
+        assertThat(diary.getTitle()).isEqualTo("오늘의 다이어리");
+        assertThat(diary.getComment()).isEqualTo("보람찬 하루였다");
```
테스트용으로 주어진 값에 오타라도 섞이면 테스트에 문제가 생길 거에요. 상수를 사용해 중복을 제거해 주세요.

johngrib reviewed (commented):
9f47e4345606cc52233e9a99fb665b6c8e5b014e 커밋 메시지에 오타가 있어요.
![image](https://user-images.githubusercontent.com/1855714/112312360-12c3f480-8cea-11eb-9e8d-523eb00c14d7.png)

johngrib commented on src/main/java/com/codesoom/project/application/DiaryService.java:
```diff
@@ -25,4 +26,20 @@
+     *
+     * @param id 다이어리 식별자
+     * @return 주어진 id를 갖는 다이어리
+     * @throws DiaryNotFoundException 주어진 id를 갖는 다이어리를 찾을 수 없을 경우
```
`다이어리를 찾을 수 없을 경우` 정도로도 충분할 것 같아요.

yhyojoo committed:
applicatino.yml 파일을 추가하라

yhyojoo committed:
하드코딩 값을 상수 값으로 변경하라

yhyojoo committed:
다이어리 서비스의 JavaDoc 주석을 수정하라

yhyojoo committed:
다이어리 서비스 테스트 코드를 계층 구조로 수정하라

yhyojoo committed:
다이어리 생성 기능을 구현하라

yhyojoo committed:
다이어리 생성 테스트를 추가하라

yhyojoo committed:
다이어리 수정 테스트를 추가하라

yhyojoo committed:
다이어리 수정 기능을 구현하라

yhyojoo committed:
다이어리 서비스 테스트 설명을 수정하라

yhyojoo committed:
다이어리 삭제 기능을 구현하고, 테스트 코드를 작성하라

yhyojoo committed:
컨트롤러에 목록 조회 기능을 구현하고 테스트 코드를 작성하라

yhyojoo committed:
다이어리 상세 조회 기능을 구현하라

yhyojoo committed:
컨트롤러 예외 처리를 위한 클래스를 생성하라

yhyojoo committed:
다이어리 상세 조회 테스트를 추가하라

yhyojoo committed:
Spring Validation 의존성을 추가하라

값에 대한 유효성 검사를 실시하기 위해 Spring Validation 의존성을 추가합니다.

yhyojoo committed:
컨트롤러 테스트의 함수명을 수정하라

yhyojoo committed:
DiaryData의 title에 제약 조건을 추가하라

yhyojoo committed:
다이어리 생성 테스트를 추가하라

yhyojoo committed:
다이어리 생성 기능을 구현하라

yhyojoo committed:
다이어리 수정 테스트를 추가하라

yhyojoo committed:
다이어리 수정 기능을 구현하라

yhyojoo committed:
다이어리 삭제 테스트를 추가하라

yhyojoo committed:
다이어리 삭제 기능을 구현하라

yhyojoo committed:
다이어리 서비스의 JavaDoc 주석을 수정하라

yhyojoo committed:
다이어리 컨트롤러 테스트를 수정하라

- 중복되는 코드를 수정합니다.
- 테스트에서 추가적으로 사용한 값을 @AfterEach에서 삭제합니다.

johngrib commented on src/main/java/com/codesoom/project/application/DiaryService.java:
```diff
@@ -25,4 +26,20 @@
+        return findDiary(id);
+    }
+
+    public Diary findDiary(Long id) {
```
여기에 JavaDoc을 작성해주면 좋겠어요.

johngrib commented on src/test/java/com/codesoom/project/controllers/DiaryControllerTest.java:
```diff
@@ -0,0 +1,359 @@
+            @DisplayName("주어진 id를 갖는 다이어리와 응답코드 200을 반환한다")
+            void it_returns_diary_and_200() throws Exception {
+                mockMvc.perform(get("/diaries/1")
+                        .accept(MediaType.APPLICATION_JSON_UTF8)
```
`MediaType.APPLICATION_JSON_UTF8`은 deprecated된 값이었죠?

johngrib commented on src/main/java/com/codesoom/project/dto/ErrorResponse.java:
```diff
@@ -0,0 +1,13 @@
+package com.codesoom.project.dto;
+
+public class ErrorResponse {
+    private String message;
```
이 변수를 `final`로 선언해도 좋겠죠.

johngrib reviewed (commented):
커밋이 많은 편이에요. 조금 더 큰 단위로 커밋해도 좋을 것 같아요.

yhyojoo committed:
JSON 응답 값을 검증하는 테스트를 추가하라

다이어리 컨트롤러 테스트에서 jsonPath 표현식을 사용해 json 응답 값을
검증합니다.

yhyojoo committed:
ErrorResponse의 message 변수를 final로 선언하라

yhyojoo committed:
ToDo 도메인 객체 클래스를 생성하라

할 일 도메인 객체 클래스를 생성하고, 빌더 테스트를 작성합니다.

yhyojoo committed:
ToDo 저장소를 생성하라

할 일 저장소를 생성하고, CRUD 메소드를 선언합니다.

yhyojoo committed:
ToDo -> Task 클래스명을 변경하라

Task가 ToDo보다 '할 일'을 더 명확히 나타내 준다고 생각되어 Task로 클래스명을 변경합니다.

yhyojoo committed:
JavaDoc 주석을 작성하라

yhyojoo committed:
할 일 목록 조회 테스트를 작성하라

할 일 목록 조회 기능에 대한 서비스와 컨트롤러 테스트를 작성합니다.

yhyojoo committed:
할 일 목록 조회 기능을 구현하라

서비스와 컨트롤러에 목록 조회 기능을 구현합니다.

yhyojoo committed:
할 일 상세 조회 테스트를 작성하라

할 일 상세 조회 기능에 대한 서비스와 컨트롤러 테스트를 작성합니다.

yhyojoo committed:
할 일 상세 조회 기능을 구현하라

서비스와 컨트롤러에 상세 조회 기능을 구현합니다.

yhyojoo committed:
할 일을 찾을 수 없는 예외를 처리하라

- 할 일을 찾을 수 없을 경우의 예외를 처리하는 클래스를 생성합니다.
- ControllerErrorAdvice 클래스에서 `@ExceptionHandler`를 이용해
  컨트롤러에서 발생하는 예외를 처리할 수 있도록 합니다.

yhyojoo committed:
JavaDoc 주석을 작성하라

yhyojoo committed:
할 일 등록 테스트를 추가하라

할 일 등록 기능에 대한 서비스와 컨트롤러 테스트를 추가합니다.

yhyojoo committed:
할 일 등록 기능을 구현하라

서비스와 컨트롤러에 할 일 등록 기능을 구현합니다.

yhyojoo reviewed (commented):
아이고 실수로 설명 고친 부분이랑 같이 커밋해 버린 것 같아요 😅
혹시 이럴 땐 본문에 추가적인 설명도 같이 적어 주어야 할까요?

yhyojoo commented:
> 아이고 실수로 설명 고친 부분이랑 같이 커밋해 버린 것 같아요 😅
> 혹시 이럴 땐 본문에 추가적인 설명도 같이 적어 주어야 할까요?

yhyojoo committed:
할 일 업데이트 기능을 구현하고 테스트를 작성하라

할 일 도메인 객체에 updateWith 메소드를 구현하고, 테스트를 작성합니다.

johngrib commented on src/test/java/com/codesoom/project/controllers/DiaryControllerTest.java:
```diff
@@ -358,5 +354,6 @@ void it_returns_404() throws Exception {
     public void afterEach() {
         diaryRepository.delete(diary);
+        diaryRepository.delete(updatedDiary);
     }
```
좋습니다. `afterEach`라는 메소드 이름은 아쉽군요. `clearContext`는 어떨까요? 이걸 `@BeforeEach`에서도 호출해주면 테스트 전에 컨텍스트를 청소해주는 용도로도 쓰게 됩니다.

johngrib reviewed (commented):
테스트코드에 대해서는 테스트코드를 가동한 결과 보고서를 스크린샷으로 찍어서 올려주시면 어떨까요? 그러면 제가 리뷰해드리기 편합니다!

yhyojoo committed:
Entity와 요청과 응답 DTO를 분리하라

지금까지는 Entity에서 모든 응답과 요청에 대한 정보를 반환했습니다. 명확한 역할 분리와 Entity 클래스의 변경으로 인해 다른 클래스에 영향을 끼치는 것을 막기 위해 정보 교환을 위한 DTO 클래스를 생성해 자주 변화가 일어나는 클래스를 Entity와 분리합니다.

yhyojoo committed:
할 일 삭제 테스트를 추가하라

할 일 삭제 기능에 대한 서비스와 컨트롤러 테스트를 추가합니다.

yhyojoo committed:
할 일 삭제 기능을 구현하라

서비스와 컨트롤러에 할 일 삭제 기능을 구현합니다.

yhyojoo committed:
테스트에서 사용한 값을 삭제하는 메소드의 이름을 변경하라

- 테스트에서 사용한 값을 지워 주는 메소드 afterEach를 메소드의 동작을 좀 더 명확히 나타내 주는 이름인 clearContext로 변경합니다.
- beforeEach에 테스트에 사용한 값을 지워 주는 코드를 추가해 테스트 전에도 컨텍스트를 청소해 줍니다.

hannut91 commented:
작업이 완료됐다면 머지 후 진행해주세요
